examples/huggingface/pytorch/text-generation/inference/run_generation_with_deepspeed.py

```diff
@@ -0,0 +1,578 @@
+        resume_download=True,
+    )
+
+
+def get_checkpoint_files(model_name_or_path):
+    cached_repo_dir = get_repo_root(model_name_or_path)
+
+    # extensions: .bin | .pt
+    # creates a list of paths from all downloaded files in cache dir
+    file_list = [str(entry) for entry in Path(cached_repo_dir).rglob("*.[bp][it][n]") if entry.is_file()]
+    return file_list
+
+
+model_name = args.model_id
+if args.dtype == "float16":
+    load_dtype = torch.half
+    infer_dtype = torch.half
+elif args.dtype == "bfloat16":
+    load_dtype = torch.bfloat16
+    infer_dtype = torch.bfloat16
+elif args.dtype == "int8":
+    load_dtype = torch.half
+    infer_dtype = torch.int8
+elif args.dtype == "float32":
+    load_dtype = torch.float32
+    infer_dtype = torch.float32
+
+tp_presharded_mode = True if model_name in tp_presharded_models else False
+
+print_rank0(f"*** Loading the model {model_name}")
+model_type = next((x for x in MODEL_CLASSES.keys() if x in model_name.lower()), 'auto')
+model_class = MODEL_CLASSES[model_type]
+tokenizer = model_class[1].from_pretrained(model_name)
+config = AutoConfig.from_pretrained(model_name, torchscript=args.jit, trust_remote_code=True)
+#if not hasattr(config, "text_max_length") and args.prompt is None:
+#    config.text_max_length = int(args.input_tokens) + int(args.max_new_tokens)
+print_rank0("*** model config:", config)
+
+# XXX: can't automatically derive dtype via config's `from_pretrained`
+# dtype = torch.bfloat16 if model_name in ["bigscience/bloom", "bigscience/bigscience-small-testing"] else torch.float16
+
+
+# use one of these args to `init_inference`
+# 1. injection_policy is the slower version, but it's plain pytorch so it'll always work
+# 2. replace_with_kernel_inject is the faster one (fast fused kernels)
+kernel_inject = args.ki
+
+if args.benchmark:
+    get_accelerator().empty_cache()
+    gc.collect()
+    deepspeed.runtime.utils.see_memory_usage("pre-from-pretrained", force=True)
+
+is_meta_support = not model_type in ["falcon"]
+
+# Construct model with fake meta tensors, later will be replaced during ds-inference ckpt load
+with deepspeed.OnDevice(dtype=load_dtype, device="meta", enabled=is_meta_support):
+    # Even inside the meta device context, from_pretrained still loads the
+    # model to cpu instead of meta device. Use from_config instead to solve the issue for big models.
+    # We add the instance type check here since some of the models haven't yet supported from_config.
+    if model_class[0] == AutoModelForCausalLM and is_meta_support:
+        model = model_class[0].from_config(config, torch_dtype=load_dtype, trust_remote_code=True)
+    else:
+        model = model_class[0].from_pretrained(model_name, config=config, low_cpu_mem_usage=True, torch_dtype=load_dtype, trust_remote_code=True)
+
+if args.benchmark:
+    deepspeed.runtime.utils.see_memory_usage("post-from-pretrained", force=True)
+
+model = model.eval()
+model = model.to(memory_format=torch.channels_last)
```

Comment:
I remember channel last is specially optimized by IPEX but not cuda. I would suggest to do this line only if ipex is found or enabled.

Reply:
I removed this.